packages/keyring-controller/src/KeyringController.ts

```diff
@@ -2077,6 +2077,51 @@ export class KeyringController extends BaseController<
     };
   }
 
+  /**
+   * Execute the given function after acquiring the controller lock
+   * and save the keyrings to state after it, or rolling back to their
+   * previous state in case of error.
+   *
+   * @param fn - The function to execute.
+   * @param options - Options for the operation.
+   * @param options.skipUpdate - Whether to skip updating the vault after the operation.
+   * @param options.onSuccess - Callback to execute after the operation is successful.
+   * @returns The result of the function.
+   */
+  async #asAtomicOperation<T>(
+    fn: MutuallyExclusiveCallback<T>,
+    options: {
+      skipUpdate?: boolean;
+      onSuccess?: () => void;
+    } = {
+      skipUpdate: false,
+    },
+  ): Promise<T> {
+    const operationResult = await this.#withControllerLock(
+      async ({ releaseLock }) => {
+        const currentKeyrings = this.#keyrings.slice();
+
+        try {
+          const callbackResult = await fn({ releaseLock });
+
+          if (!options.skipUpdate) {
+            await this.#updateVault();
+          }
+
+          return callbackResult;
+        } catch (e) {
+          // We rollback the keyrings to the previous state
+          this.#keyrings = currentKeyrings;
+          throw e;
+        }
+      },
+    );
+
+    options.onSuccess?.();
```

Comment:
We could also execute this `onSuccess` hook inside the lock, but calling it after releasing gives a way to publish messenger events without risking deadlocks (if an event listener calls some other function of this controller)

Reply:
It also allows methods like `removeAccount` to emit an event after `this.#updateVault` has been called internally by `#asAtomicOperation`

Reply:
Hmm. Putting it after releasing the lock is equivalent to just make calls in the wrapper caller after it's resolved. We could just remove it in this case